File: windows/IDPersonalSecure/Data/Vault.cs
using System.Collections.ObjectModel;
using System.IO;
using System.IO.Compression;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using IDPersonalSecure.Crypto;

namespace IDPersonalSecure.Data;

public sealed class IntegrityException(string message) : Exception(message);

public sealed record DocType(string Code, string Label, string Country);

public static class DocumentCatalog
{
    public static readonly IReadOnlyList<DocType> Types = new List<DocType>
    {
        new("REGISTRO_CIVIL", "Registro Civil", "CO"),
        new("TI", "Tarjeta de Identidad", "CO"),
        new("CC", "Cédula de Ciudadanía", "CO"),
        new("CE", "Cédula de Extranjería", "CO"),
        new("PASAPORTE", "Pasaporte", "CO"),
        new("PPT", "Permiso por Protección Temporal", "CO"),
        new("SC", "Salvoconducto", "CO"),
        new("NIT", "NIT", "CO"),
        new("DNI", "DNI", "XX"),
        new("CURP", "CURP", "MX"),
        new("RUT", "RUT", "CL"),
        new("PASSPORT", "Passport", "XX"),
        new("CERT", "Certificado", "XX"),
    };

    public static string Label(string code) =>
        Types.FirstOrDefault(t => t.Code == code)?.Label ?? code;
}

public sealed class Document
{
    public string Id { get; set; } = Guid.NewGuid().ToString();
    public string Name { get; set; } = "";
    public string Type { get; set; } = "CC";
    public string Country { get; set; } = "CO";
    public string Number { get; set; } = "";
    public string IssueDate { get; set; } = "";
    public string ExpiryDate { get; set; } = "";
    public bool HasExpiry { get; set; }
    public string UrlSource { get; set; } = "";
    public string FileName { get; set; } = "";
    public string Notes { get; set; } = "";
    public string ReminderAt { get; set; } = "";
    public string CreatedAt { get; set; } = DateTimeOffset.UtcNow.ToString("yyyy-MM-ddTHH:mm:ssZ");

    [JsonIgnore] public string TypeLabel => DocumentCatalog.Label(Type);
    [JsonIgnore] public string AttachmentBadge => string.IsNullOrEmpty(FileName) ? "" : "📎";
    [JsonIgnore] public bool HasUrl => !string.IsNullOrWhiteSpace(UrlSource);
    [JsonIgnore] public string Subtitle => $"{TypeLabel} · {Country}" + (string.IsNullOrEmpty(Number) ? "" : $" · N.º {Number}");
    [JsonIgnore] public bool IsExpired => HasExpiry && DateTime.TryParse(ExpiryDate, out var d) && d.Date < DateTime.Today;
    [JsonIgnore] public string ExpiryDisplay =>
        HasExpiry && !string.IsNullOrEmpty(ExpiryDate) ? (IsExpired ? $"Vencido: {ExpiryDate}" : $"Vence: {ExpiryDate}") : "";

    public Document Clone() => (Document)MemberwiseClone();
}

/// <summary>Registro de un documento compartido (historial interno).</summary>
public sealed class ShareRecord
{
    public string Id { get; set; } = "";
    public string DocId { get; set; } = "";
    public string DocName { get; set; } = "";
    public string Tramite { get; set; } = "";
    public string DateTime { get; set; } = "";
    public string Recipient { get; set; } = "";
    public bool Watermarked { get; set; }

    [JsonIgnore] public string Line1 => $"{DocName}  ·  {DateTime}";
    [JsonIgnore] public string Line2 => (Watermarked ? $"ID {Id}" : "sin marca de agua") +
        (string.IsNullOrWhiteSpace(Tramite) ? "" : $"  ·  {Tramite}");
    [JsonIgnore] public string RecipientDisplay => string.IsNullOrWhiteSpace(Recipient) ? "Destinatario: —" : $"Destinatario: {Recipient}";
}

public sealed class VaultDb
{
    public int Schema { get; set; } = 1;
    public List<Document> Documents { get; set; } = new();
    public List<ShareRecord> ShareLog { get; set; } = new();
}

file sealed class Manifest
{
    public string Format { get; set; } = "securevault";
    public int FormatVersion { get; set; } = 1;
    public string App { get; set; } = "IDPersonalSecure";
    public string CreatedAt { get; set; } = "";
    public string Kdf { get; set; } = "PBKDF2-HMAC-SHA256";
    public int Iterations { get; set; } = 210_000;
    public string Cipher { get; set; } = "AES-256-GCM";
    public string Salt { get; set; } = "";
    public string DbMac { get; set; } = "";
    public List<string> Files { get; set; } = new();
}

/// <summary>
/// Bóveda local cifrada + export/import de `.securevault`.
/// Mismo esquema que Android (docs/CRYPTO.md) para interoperabilidad directa.
/// </summary>
public sealed class VaultRepository
{
    private static readonly Encoding Utf8NoBom = new UTF8Encoding(false);
    private static readonly JsonSerializerOptions Json = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        PropertyNameCaseInsensitive = true,
    };

    private readonly string _dir;
    private string DbPath => Path.Combine(_dir, "vault.db.enc");
    private string SaltPath => Path.Combine(_dir, "vault.salt");
    private string AttachDir => Path.Combine(_dir, "attachments");
    private string AttachPath(string id) => Path.Combine(AttachDir, $"{id}.enc");

    public byte[] Salt { get; private set; } = Array.Empty<byte>();
    private VaultCrypto.Keys? _keys;
    public ObservableCollection<Document> Documents { get; } = new();
    public ObservableCollection<ShareRecord> ShareLog { get; } = new();

    public VaultRepository()
    {
        _dir = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "IDPersonalSecure");
        Directory.CreateDirectory(_dir);
        Directory.CreateDirectory(AttachDir);
    }

    // ── Adjuntos cifrados (files/<id>.enc, AAD = id) ─────────────────────
    public bool HasAttachment(string id) => File.Exists(AttachPath(id));

    public void SaveAttachment(string id, byte[] data)
    {
        if (_keys is null) throw new InvalidOperationException("Bóveda bloqueada");
        Directory.CreateDirectory(AttachDir);
        File.WriteAllBytes(AttachPath(id), VaultCrypto.Encrypt(data, _keys.Enc, id));
    }

    public byte[]? ReadAttachment(string id)
    {
        if (_keys is null || !HasAttachment(id)) return null;
        return VaultCrypto.Decrypt(File.ReadAllBytes(AttachPath(id)), _keys.Enc, id);
    }

    public void DeleteAttachment(string id)
    {
        if (File.Exists(AttachPath(id))) File.Delete(AttachPath(id));
    }

    public bool VaultExists() => File.Exists(SaltPath);
    public bool IsUnlocked => _keys is not null;

    public bool Unlock(string pin)
    {
        Salt = File.Exists(SaltPath) ? File.ReadAllBytes(SaltPath) : CreateSalt();
        var k = VaultCrypto.DeriveKeys(pin, Salt);
        Documents.Clear();
        if (File.Exists(DbPath))
        {
            try
            {
                string json = Encoding.UTF8.GetString(VaultCrypto.Decrypt(File.ReadAllBytes(DbPath), k.Enc, "database"));
                LoadJson(json);
            }
            catch
            {
                return false; // PIN incorrecto o datos corruptos
            }
        }
        _keys = k;
        return true;
    }

    public void Lock() { _keys = null; Documents.Clear(); }

    private byte[] CreateSalt()
    {
        var s = VaultCrypto.NewSalt();
        File.WriteAllBytes(SaltPath, s);
        return s;
    }

    private void LoadJson(string json)
    {
        Documents.Clear();
        ShareLog.Clear();
        var db = JsonSerializer.Deserialize<VaultDb>(json, Json) ?? new VaultDb();
        foreach (var d in db.Documents) Documents.Add(d);
        foreach (var s in db.ShareLog) ShareLog.Add(s);
    }

    private byte[] SerializeDb() =>
        Encoding.UTF8.GetBytes(JsonSerializer.Serialize(
            new VaultDb { Documents = Documents.ToList(), ShareLog = ShareLog.ToList() }, Json));

    public void AddShareRecord(ShareRecord rec) { ShareLog.Insert(0, rec); Save(); }

    public void UpdateShareRecipient(string id, string recipient)
    {
        var r = ShareLog.FirstOrDefault(x => x.Id == id);
        if (r != null) { r.Recipient = recipient; Save(); }
    }

    public void Save()
    {
        if (_keys is null) return;
        File.WriteAllBytes(DbPath, VaultCrypto.Encrypt(SerializeDb(), _keys.Enc, "database"));
    }

    public void Upsert(Document doc)
    {
        var idx = Documents.ToList().FindIndex(d => d.Id == doc.Id);
        if (idx >= 0) Documents[idx] = doc; else Documents.Add(doc);
        Save();
    }

    public void Delete(string id)
    {
        var existing = Documents.FirstOrDefault(d => d.Id == id);
        if (existing != null) Documents.Remove(existing);
        DeleteAttachment(id);
        Save();
    }

    public void Export(Stream outStream)
    {
        if (_keys is null) throw new InvalidOperationException("Bóveda bloqueada");
        var dbBlob = VaultCrypto.Encrypt(SerializeDb(), _keys.Enc, "database");
        string dbB64 = VaultCrypto.B64(dbBlob);
        string dbMac = VaultCrypto.B64(VaultCrypto.Hmac(_keys.Mac, Encoding.UTF8.GetBytes(dbB64)));
        var manifest = new Manifest
        {
            CreatedAt = DateTimeOffset.UtcNow.ToString("yyyy-MM-ddTHH:mm:ssZ"),
            Salt = VaultCrypto.B64(Salt),
            DbMac = dbMac,
        };
        var fileIds = Documents.Where(d => HasAttachment(d.Id)).Select(d => d.Id).ToList();
        manifest.Files = fileIds;

        using var zip = new ZipArchive(outStream, ZipArchiveMode.Create, leaveOpen: true);
        using (var w = new StreamWriter(zip.CreateEntry("manifest.json").Open(), Utf8NoBom))
            w.Write(JsonSerializer.Serialize(manifest, new JsonSerializerOptions(Json) { WriteIndented = true }));
        using (var w = new StreamWriter(zip.CreateEntry("database.enc").Open(), Utf8NoBom))
            w.Write(dbB64);
        foreach (var id in fileIds)
        {
            string blobB64 = VaultCrypto.B64(File.ReadAllBytes(AttachPath(id)));
            using var w = new StreamWriter(zip.CreateEntry($"files/{id}.enc").Open(), Utf8NoBom);
            w.Write(blobB64);
        }
    }

    public void Import(Stream input, string pin)
    {
        string? manifestText = null, dbB64 = null;
        var fileBlobs = new Dictionary<string, string>();
        using (var zip = new ZipArchive(input, ZipArchiveMode.Read, leaveOpen: true))
        {
            foreach (var e in zip.Entries)
            {
                using var r = new StreamReader(e.Open(), Encoding.UTF8);
                string content = r.ReadToEnd();
                if (e.FullName == "manifest.json") manifestText = content;
                else if (e.FullName == "database.enc") dbB64 = content;
                else if (e.FullName.StartsWith("files/") && e.FullName.EndsWith(".enc"))
                    fileBlobs[Path.GetFileNameWithoutExtension(e.Name)] = content;
            }
        }
        if (manifestText is null) throw new IntegrityException("manifest.json ausente");
        if (dbB64 is null) throw new IntegrityException("database.enc ausente");

        using var doc = JsonDocument.Parse(manifestText);
        var root = doc.RootElement;
        byte[] impSalt = VaultCrypto.Unb64(root.GetProperty("salt").GetString()!);
        string expectMac = root.GetProperty("dbMac").GetString()!;
        var k = VaultCrypto.DeriveKeys(pin, impSalt);
        string actualMac = VaultCrypto.B64(VaultCrypto.Hmac(k.Mac, Encoding.UTF8.GetBytes(dbB64)));
        if (!VaultCrypto.ConstantTimeEquals(expectMac, actualMac))
            throw new IntegrityException("Integridad inválida o PIN incorrecto");

        string json = Encoding.UTF8.GetString(VaultCrypto.Decrypt(VaultCrypto.Unb64(dbB64), k.Enc, "database"));
        Salt = impSalt;
        File.WriteAllBytes(SaltPath, impSalt);
        _keys = k;
        LoadJson(json);
        Save();

        // Restaura adjuntos: limpia los previos y escribe los importados (misma clave).
        if (Directory.Exists(AttachDir))
            foreach (var f in Directory.GetFiles(AttachDir)) File.Delete(f);
        Directory.CreateDirectory(AttachDir);
        foreach (var (id, b64) in fileBlobs)
            File.WriteAllBytes(AttachPath(id), VaultCrypto.Unb64(b64));
    }
}

File: android/app/src/main/java/com/idpersonalsecure/app/data/Vault.kt
package com.idpersonalsecure.app.data

import android.content.Context
import com.idpersonalsecure.app.crypto.VaultCrypto
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.InputStream
import java.io.OutputStream
import java.time.OffsetDateTime
import java.time.ZoneOffset
import java.util.UUID
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import java.util.zip.ZipOutputStream

fun nowIso(): String = OffsetDateTime.now(ZoneOffset.UTC).withNano(0).toString()

class IntegrityException(msg: String) : Exception(msg)

/** Tipo de documento del catálogo (Colombia + multipaís). */
data class DocType(val code: String, val label: String, val country: String)

object DocumentCatalog {
    val types: List<DocType> = listOf(
        DocType("REGISTRO_CIVIL", "Registro Civil", "CO"),
        DocType("TI", "Tarjeta de Identidad", "CO"),
        DocType("CC", "Cédula de Ciudadanía", "CO"),
        DocType("CE", "Cédula de Extranjería", "CO"),
        DocType("PASAPORTE", "Pasaporte", "CO"),
        DocType("PPT", "Permiso por Protección Temporal", "CO"),
        DocType("SC", "Salvoconducto", "CO"),
        DocType("NIT", "NIT", "CO"),
        DocType("DNI", "DNI", "XX"),
        DocType("CURP", "CURP", "MX"),
        DocType("RUT", "RUT", "CL"),
        DocType("PASSPORT", "Passport", "XX"),
        DocType("CERT", "Certificado", "XX"),
    )
    fun label(code: String): String = types.firstOrNull { it.code == code }?.label ?: code
}

data class Document(
    val id: String = UUID.randomUUID().toString(),
    var name: String = "",
    var type: String = "CC",
    var country: String = "CO",
    var number: String = "",
    var issueDate: String = "",
    var expiryDate: String = "",
    var hasExpiry: Boolean = false,
    var urlSource: String = "",
    var fileName: String = "",
    var notes: String = "",
    val createdAt: String = nowIso(),
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("id", id); put("name", name); put("type", type); put("country", country)
        put("number", number); put("issueDate", issueDate); put("expiryDate", expiryDate)
        put("hasExpiry", hasExpiry); put("urlSource", urlSource); put("fileName", fileName)
        put("notes", notes); put("createdAt", createdAt)
    }

    companion object {
        fun fromJson(o: JSONObject) = Document(
            id = o.optString("id", UUID.randomUUID().toString()),
            name = o.optString("name"),
            type = o.optString("type", "CC"),
            country = o.optString("country", "CO"),
            number = o.optString("number"),
            issueDate = o.optString("issueDate"),
            expiryDate = o.optString("expiryDate"),
            hasExpiry = o.optBoolean("hasExpiry", false),
            urlSource = o.optString("urlSource"),
            fileName = o.optString("fileName"),
            notes = o.optString("notes"),
            createdAt = o.optString("createdAt", nowIso()),
        )
    }
}

/** Registro de un documento compartido (historial interno). */
data class ShareRecord(
    val id: String,
    val docId: String,
    val docName: String,
    val tramite: String,
    val dateTime: String,
    var recipient: String,
    val watermarked: Boolean,
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("id", id); put("docId", docId); put("docName", docName); put("tramite", tramite)
        put("dateTime", dateTime); put("recipient", recipient); put("watermarked", watermarked)
    }

    companion object {
        fun fromJson(o: JSONObject) = ShareRecord(
            id = o.optString("id"), docId = o.optString("docId"), docName = o.optString("docName"),
            tramite = o.optString("tramite"), dateTime = o.optString("dateTime"),
            recipient = o.optString("recipient"), watermarked = o.optBoolean("watermarked", false),
        )
    }
}

/**
 * Bóveda local cifrada + export/import de `.securevault`.
 * El almacén local usa el MISMO esquema que el archivo exportado (docs/CRYPTO.md).
 */
class VaultRepository(context: Context) {
    private val filesDir: File = context.filesDir
    private val dbFile get() = File(filesDir, "vault.db.enc")
    private val saltFile get() = File(filesDir, "vault.salt")
    private val attachDir get() = File(filesDir, "attachments").apply { mkdirs() }
    private fun attachFile(id: String) = File(attachDir, "$id.enc")

    var salt: ByteArray = ByteArray(0); private set
    private var keys: VaultCrypto.Keys? = null
    val documents = mutableListOf<Document>()
    val shareLog = mutableListOf<ShareRecord>()

    fun vaultExists(): Boolean = saltFile.exists()
    val isUnlocked: Boolean get() = keys != null

    /** Desbloquea (o crea) la bóveda local con el PIN. false si el PIN es incorrecto. */
    fun unlock(pin: String): Boolean {
        salt = if (saltFile.exists()) saltFile.readBytes()
        else VaultCrypto.newSalt().also { saltFile.writeBytes(it) }
        val k = VaultCrypto.deriveKeys(pin, salt)
        documents.clear()
        if (dbFile.exists()) {
            try {
                loadJson(String(VaultCrypto.decrypt(dbFile.readBytes(), k.enc, "database"), Charsets.UTF_8))
            } catch (e: Exception) {
                return false
            }
        }
        keys = k
        return true
    }

    fun lock() { keys = null; documents.clear() }

    private fun loadJson(json: String) {
        documents.clear()
        shareLog.clear()
        val root = JSONObject(json)
        val arr = root.optJSONArray("documents") ?: JSONArray()
        for (i in 0 until arr.length()) documents.add(Document.fromJson(arr.getJSONObject(i)))
        val log = root.optJSONArray("shareLog") ?: JSONArray()
        for (i in 0 until log.length()) shareLog.add(ShareRecord.fromJson(log.getJSONObject(i)))
    }

    private fun buildDbJson(): String {
        val arr = JSONArray()
        documents.forEach { arr.put(it.toJson()) }
        val log = JSONArray()
        shareLog.forEach { log.put(it.toJson()) }
        return JSONObject().put("schema", 1).put("documents", arr).put("shareLog", log).toString()
    }

    fun addShareRecord(rec: ShareRecord) { shareLog.add(0, rec); save() }

    fun updateShareRecipient(id: String, recipient: String) {
        shareLog.firstOrNull { it.id == id }?.let { it.recipient = recipient; save() }
    }

    fun save() {
        val k = keys ?: return
        dbFile.writeBytes(VaultCrypto.encrypt(buildDbJson().toByteArray(Charsets.UTF_8), k.enc, "database"))
    }

    fun upsert(doc: Document) {
        val idx = documents.indexOfFirst { it.id == doc.id }
        if (idx >= 0) documents[idx] = doc else documents.add(doc)
        save()
    }

    fun delete(id: String) { documents.removeAll { it.id == id }; deleteAttachment(id); save() }

    // ── Adjuntos cifrados (files/<id>.enc, AAD = id) ─────────────────────
    fun hasAttachment(id: String): Boolean = attachFile(id).exists()

    fun saveAttachment(id: String, data: ByteArray) {
        val k = keys ?: throw IllegalStateException("Bóveda bloqueada")
        attachFile(id).writeBytes(VaultCrypto.encrypt(data, k.enc, id))
    }

    fun readAttachment(id: String): ByteArray? {
        val k = keys ?: return null
        val f = attachFile(id)
        return if (f.exists()) VaultCrypto.decrypt(f.readBytes(), k.enc, id) else null
    }

    fun deleteAttachment(id: String) { attachFile(id).delete() }

    /** Exporta la bóveda actual a un `.securevault`. */
    fun export(out: OutputStream) {
        val k = keys ?: throw IllegalStateException("Bóveda bloqueada")
        val dbBlob = VaultCrypto.encrypt(buildDbJson().toByteArray(Charsets.UTF_8), k.enc, "database")
        val dbB64 = VaultCrypto.b64(dbBlob)
        val dbMac = VaultCrypto.b64(VaultCrypto.hmac(k.mac, dbB64.toByteArray(Charsets.UTF_8)))
        val fileIds = documents.filter { hasAttachment(it.id) }.map { it.id }
        val manifest = JSONObject().apply {
            put("format", "securevault"); put("formatVersion", 1); put("app", "IDPersonalSecure")
            put("createdAt", nowIso()); put("kdf", "PBKDF2-HMAC-SHA256"); put("iterations", 210000)
            put("cipher", "AES-256-GCM"); put("salt", VaultCrypto.b64(salt)); put("dbMac", dbMac)
            put("files", JSONArray(fileIds))
        }
        ZipOutputStream(out).use { zip ->
            zip.putNextEntry(ZipEntry("manifest.json"))
            zip.write(manifest.toString(2).toByteArray(Charsets.UTF_8)); zip.closeEntry()
            zip.putNextEntry(ZipEntry("database.enc"))
            zip.write(dbB64.toByteArray(Charsets.UTF_8)); zip.closeEntry()
            for (id in fileIds) {
                val blobB64 = VaultCrypto.b64(attachFile(id).readBytes())
                zip.putNextEntry(ZipEntry("files/$id.enc"))
                zip.write(blobB64.toByteArray(Charsets.UTF_8)); zip.closeEntry()
            }
        }
    }

    /** Importa un `.securevault`: valida integridad (HMAC) y reemplaza la bóveda local. */
    fun import(input: InputStream, pin: String) {
        var manifestText: String? = null
        var dbB64: String? = null
        val fileBlobs = HashMap<String, String>()
        ZipInputStream(input).use { zip ->
            var e: ZipEntry? = zip.nextEntry
            while (e != null) {
                val bytes = readAll(zip)
                val name = e.name
                when {
                    name == "manifest.json" -> manifestText = String(bytes, Charsets.UTF_8)
                    name == "database.enc" -> dbB64 = String(bytes, Charsets.UTF_8)
                    name.startsWith("files/") && name.endsWith(".enc") ->
                        fileBlobs[name.substringAfterLast('/').removeSuffix(".enc")] = String(bytes, Charsets.UTF_8)
                }
                zip.closeEntry(); e = zip.nextEntry
            }
        }
        val manifest = JSONObject(manifestText ?: throw IntegrityException("manifest.json ausente"))
        val impSalt = VaultCrypto.unb64(manifest.getString("salt"))
        val db = dbB64 ?: throw IntegrityException("database.enc ausente")
        val k = VaultCrypto.deriveKeys(pin, impSalt)
        val expectMac = manifest.getString("dbMac")
        val actualMac = VaultCrypto.b64(VaultCrypto.hmac(k.mac, db.toByteArray(Charsets.UTF_8)))
        if (!constantTimeEquals(expectMac, actualMac)) throw IntegrityException("Integridad inválida o PIN incorrecto")
        val json = String(VaultCrypto.decrypt(VaultCrypto.unb64(db), k.enc, "database"), Charsets.UTF_8)
        salt = impSalt; saltFile.writeBytes(impSalt); keys = k
        loadJson(json); save()

        // Restaura adjuntos: limpia los previos y escribe los importados (misma clave).
        attachDir.listFiles()?.forEach { it.delete() }
        for ((id, b64) in fileBlobs) attachFile(id).writeBytes(VaultCrypto.unb64(b64))
    }

    private fun constantTimeEquals(a: String, b: String): Boolean {
        val x = a.toByteArray(); val y = b.toByteArray()
        if (x.size != y.size) return false
        var r = 0
        for (i in x.indices) r = r or (x[i].toInt() xor y[i].toInt())
        return r == 0
    }

    private fun readAll(input: InputStream): ByteArray {
        val bos = ByteArrayOutputStream(); val buf = ByteArray(8192); var n: Int
        while (input.read(buf).also { n = it } >= 0) bos.write(buf, 0, n)
        return bos.toByteArray()
    }
}
